### PR TITLE
feat: support concurrent installation of browsers

### LINKED ISSUE #3912: [BUG] Race condition with concurrent playwright NPM postinstall scripts
**Context:**
- Playwright Version: 1.4.0
- Operating System: macOS 10.15.6
- Node.js version: 14.6.0
- Browser: N/A
- Extra: N/A

**Code Snippet**

```bash
export PLAYWRIGHT_BROWSERS_PATH=~/playwright
rm -rf $PLAYWRIGHT_BROWSERS_PATH
node node_modules/playwright/install.js & node node_modules/playwright/install.js
```

Output:

```
Downloading chromium v799411 - 118.1 Mb [====================] 100% 0.0s 
Downloading chromium v799411 - 118.1 Mb [====================] 100% 0.0s 
(node:22246) UnhandledPromiseRejectionWarning: Error: EEXIST: file already exists, symlink 'Versions/Current/Chromium Framework' -> '/Users/aomarks/playwright/chromium-799411/chrome-mac/Chromium.app/Contents/Frameworks/Chromium Framework.framework/Chromium Framework'
(Use `node --trace-warnings ...` to show where the warning was created)
(node:22246) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:22246) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Describe the bug**

If two `npm installs` are happening on the same system at the same time, and they both depend on the `playwright` package, then it is possible for multiple invocations of the `playwright` postinstall script (which downloads the browser binaries to a common system location) to be running at the same time. It seems this can cause a race condition due to both instances are downloading the same browser to the same location, which will cause one of the processes to crash.

One way this can come up in practice is in a Lerna monorepo, where two sub-packages both depend on `playwright`, because the `lerna bootstrap` command runs multiple `npm installs` concurrently. (Our workaround for the time being is to run `lerna bootstrap --concurrency=1`).

### PR DESCRIPTION
**Pull request by aslushnikov**
A few details on locking registry to prohibit concurrent access:
- locking is done by creating a `__dirlock` directory in the top-level
of our registry.
- since `__dirlock` directory does not match any of browser
directories, old versions of the installer will ignore it
- in case of concurrent access, installation will wait for a lock to be
released for 10 minutes, periodically trying to grab the lock. If it
fails to do so in 10 minutes, the installation will fail.

Fixes #3912

### CODE REVIEW

**Review by yury-s — APPROVED** [review]

File: src/install/installer.ts [hunk 1]
```diff
@@ -33,6 +34,17 @@ const removeFolderAsync = util.promisify(removeFolder);
...
 export async function installBrowsersWithProgressBar(packagePath: string) {
   const browsersPath = browserPaths.browsersPath(packagePath);
+  await fsMkdirAsync(browsersPath, { recursive: true });
+  const releaseLock = await lockfile.lock(browsersPath, {
+    retries: {
+      retries: 10,
+      // Retry 20 times during 10 minutes with
+      // exponential back-off.
+      // See documentation at: https://www.npmjs.com/package/retry#retrytimeoutsoptions
+      factor: 1.27579,
```
Comment (mxschmitt) [inline src/install/installer.ts:44]: Shouldn't the comment be "Retry 10 times..." and the factor can be simplified to `1.26`?

Comment (mxschmitt) [inline src/install/installer.ts:44]: Shouldn't the comment be "Retry 10 times..." and the factor can be simplified to `1.26`?

Comment (aslushnikov) [reply src/install/installer.ts:44]: @mxschmitt I think it is exactly what it is - check out this wolframalpha link: https://www.wolframalpha.com/input/?i=Sum%5B1000*x%5Ek%2C+%7Bk%2C+0%2C+20%7D%5D+%3D+10+*+60+*+1000

Comment (mxschmitt) [reply src/install/installer.ts:44]: okay true.

**Pull request merged**